lib/Sema/CSDiag.cpp

```diff
@@ -5160,12 +5160,12 @@ bool FailureDiagnosis::diagnoseTrailingClosureErrors(ApplyExpr *callExpr) {
 
   SmallVector<Type, 4> possibleTypes;
   auto currentType = CS.getType(fnExpr);
+  auto contextualType = CS.getContextualType();
 
   // If current type has type variables or unresolved types
   // let's try to re-typecheck it to see if we can get some
   // more information about what is going on.
-  if (currentType->hasTypeVariable() || currentType->hasUnresolvedType()) {
-    auto contextualType = CS.getContextualType();
+  if (!currentType->hasTypeVariable() && currentType->hasUnresolvedType()) {
```

Comment:
As far as I recall the reason why it was like this is that by type-checking separately we could get a lot more information coming from the argument itself (if this is a contextual problem), so we don't really want to shy away from re-checking unresolved or type var containing things, since they should be erased.